packages/playwright/src/runner/watchMode.ts

```diff
@@ -250,71 +237,36 @@ export async function runWatchModeLoop(config: FullConfigInternal): Promise<Full
     }
   }
 
-  const cleanupStatus = await globalCleanup();
-  return result === 'passed' ? cleanupStatus : result;
-}
+  const teardown = await testServerConnection.runGlobalTeardown({});
 
-async function runChangedTests(config: FullConfigInternal, failedTestIdCollector: Set<string>, filesByProject: Map<FullProjectInternal, Set<string>>, title?: string) {
-  const testFiles = new Set<string>();
-  for (const files of filesByProject.values())
-    files.forEach(f => testFiles.add(f));
-
-  // Collect all the affected projects, follow project dependencies.
-  // Prepare to exclude all the projects that do not depend on this file, as if they did not exist.
-  const projects = filterProjects(config.projects, config.cliProjectFilter);
-  const projectClosure = buildProjectsClosure(projects);
-  const affectedProjects = affectedProjectsClosure([...projectClosure.keys()], [...filesByProject.keys()]);
-  const affectsAnyDependency = [...affectedProjects].some(p => projectClosure.get(p) === 'dependency');
-
-  // If there are affected dependency projects, do the full run, respect the original CLI.
-  // if there are no affected dependency projects, intersect CLI with dirty files
-  const additionalFileMatcher = affectsAnyDependency ? () => true : (file: string) => testFiles.has(file);
-  await runTests(config, failedTestIdCollector, { additionalFileMatcher, title: title || 'files changed' });
+  return result === 'passed' ? teardown.status : result;
 }
 
-async function runTests(config: FullConfigInternal, failedTestIdCollector: Set<string>, options?: {
-    projectsToIgnore?: Set<FullProjectInternal>,
-    additionalFileMatcher?: Matcher,
-    title?: string,
-  }) {
-  printConfiguration(config, options?.title);
-  const taskRunner = createTaskRunnerForWatch(config, [new ListReporter()], options?.additionalFileMatcher);
-  const testRun = new TestRun(config);
-  taskRunner.reporter.onConfigure(config.config);
-  const taskStatus = await taskRunner.run(testRun, 0);
-  let status: FullResult['status'] = 'passed';
-
-  let hasFailedTests = false;
-  for (const test of testRun.rootSuite?.allTests() || []) {
-    if (test.outcome() === 'unexpected') {
-      failedTestIdCollector.add(test.id);
-      hasFailedTests = true;
-    } else {
-      failedTestIdCollector.delete(test.id);
-    }
-  }
 
-  if (testRun.failureTracker.hasWorkerErrors() || hasFailedTests)
-    status = 'failed';
-  if (status === 'passed' && taskStatus !== 'passed')
-    status = taskStatus;
-  await taskRunner.reporter.onEnd({ status });
-  await taskRunner.reporter.onExit();
+async function runChangedTests(watchOptions: WatchModeOptions, testServerConnection: TestServerConnection, changedFiles: string[], title?: string) {
+  if (watchOptions.files?.length)
+    changedFiles = changedFiles.filter(createFileMatcherFromArguments(watchOptions.files));
+
+  await runTests(watchOptions, testServerConnection, { title: title || 'files changed', locations: changedFiles });
```

Comment:
I think it is now easier to pass `testId` filter instead, given that you have `teleSuiteUpdater.rootSuite.allTests()`.